lib/pure/endians.nim

```diff
@@ -98,16 +135,853 @@ else:
+    copyMem(outp, inp, 8)
+
+  proc littleEndian32*(outp, inp: pointer) {.inline, deprecated:
+      "Deprecated since v1.1.0; use `fromBytesXX/toBytesXX` instead".} =
+    copyMem(outp, inp, 4)
+
+  proc littleEndian16*(outp, inp: pointer){.inline, deprecated:
+      "Deprecated since v1.1.0; use `fromBytesXX/toBytesXX` instead".} =
+    copyMem(outp, inp, 2)
+
+  proc bigEndian64*(outp, inp: pointer) {.inline, deprecated:
+      "Deprecated since v1.1.0; use `fromBytesXX/toBytesXX` instead".} =
+    swapEndian64(outp, inp)
+
+  proc bigEndian32*(outp, inp: pointer) {.inline, deprecated:
+      "Deprecated since v1.1.0; use `fromBytesXX/toBytesXX` instead".} =
+    swapEndian32(outp, inp)
+
+  proc bigEndian16*(outp, inp: pointer) {.inline, deprecated:
+      "Deprecated since v1.1.0; use `fromBytesXX/toBytesXX` instead".} =
+    swapEndian16(outp, inp)
+
+
+# New API starts here
+
+func slowSwap16(a: uint16): uint16 {.inline.} =
+  result = (a shl 8) or (a shr 8)
+
+func slowSwap32(a: uint32): uint32 {.inline.} =
+  var a = ((a shl  8) and 0xff00ff00'u32) or
+          ((a shr  8) and 0x00ff00ff'u32)
+  result = (a shl 16) or (a shr 16)
+
+func slowSwap64(a: uint64): uint64 {.inline.} =
+  var a = ((a shl 8) and 0xff00ff00ff00ff00'u64) or
+          ((a shr 8) and 0x00ff00ff00ff00ff'u64)
+  a = ((a shl 16) and 0xffff0000ffff0000'u64) or
+      ((a shr 16) and 0x0000ffff0000ffff'u64)
+  result = (a shl 32) or (a shr 32)
+
+when useBuiltinSwap:
+  template swap16(a: uint16): uint16 = builtinSwap16(a)
+  template swap32(a: uint32): uint32 = builtinSwap32(a)
+  template swap64(a: uint64): uint64 = builtinSwap64(a)
 else:
-  proc littleEndian64*(outp, inp: pointer) {.inline.} = copyMem(outp, inp, 8)
-  proc littleEndian32*(outp, inp: pointer) {.inline.} = copyMem(outp, inp, 4)
-  proc littleEndian16*(outp, inp: pointer){.inline.} = copyMem(outp, inp, 2)
-  proc bigEndian64*(outp, inp: pointer) {.inline.} = swapEndian64(outp, inp)
-  proc bigEndian32*(outp, inp: pointer) {.inline.} = swapEndian32(outp, inp)
-  proc bigEndian16*(outp, inp: pointer) {.inline.} = swapEndian16(outp, inp)
+  template swap16(a: uint16): uint16 = slowSwap16(a)
+  template swap32(a: uint32): uint32 = slowSwap32(a)
+  template swap64(a: uint64): uint64 = slowSwap64(a)
+
+
+func swapEndian*[T: SomeNumber](value: T): T {.inline, mySince: (1,1).} =
+  ## Swaps the byte order of a number.
+  runnableExamples:
+    doAssert swapEndian(0xdeadbeef'i32) == 0xefbeadde'i32
+
+  when T is SomeInteger8:
+    value
+  elif T is SomeInteger16:
+    when nimvm: cast[T](slowSwap16(cast[uint16](value)))
+    else:       cast[T](    swap16(cast[uint16](value)))
+  elif T is SomeNumber32:
+    when nimvm: cast[T](slowSwap32(cast[uint32](value)))
+    else:       cast[T](    swap32(cast[uint32](value)))
+  elif T is SomeNumber64:
+    when nimvm: cast[T](slowSwap64(cast[uint64](value)))
+    else:       cast[T](    swap64(cast[uint64](value)))
+
+
+proc swapEndian*(T: typedesc[SomeNumber],
```

Comment:
Can these be funcs too?

Reply:
They have side effects.

Reply:
Wait what?

Reply:
Okay, I tried changing it to `func` and it still works. Hmmm...

This was a surprise to me as I'm doing pure FP-style programming as my day job where mutating passed in objects or changing random memory locations is definitely considered a side effect! The standard functional programming definition for side effects is that the function must have no observable effects apart from returning a value.

Anyway, you guys have redefined this term a bit, but that's fine, I just wasn't aware of it.

Reply:
Okay, just realised that when you guys say "side effects", you don't actually mean the [standard definition](https://en.wikipedia.org/wiki/Side_effect_(computer_science)) :) I do pure-FP style coding as my day job in Scala, so this surprised me a bit. But okay... in the sense how Nim views side effects, these are side effect free (they don't touch any globals). Made them all funcs.